### ISSUE TITLE
Use Probot to Manage Branch Protection Rules as Code

### ISSUE THREAD

**Issue by pjhill** [department-of-veterans-affairs/va.gov-team#56437]
# Description
Given that the GitHub branch protection rules are currently only defined via the UI, and access to the repository settings is rightfully restricted to few people -- the organization is currently unable to support restoration of the branch protection rules in the event of a disaster.

Instead, how might we work to ensure that branch protection rules are defined in code and applied from a Settings app ([like this one](https://github.com/apps/settings)) potentially using automation like [Probot](https://github.com/probot/probot).

# Acceptance Criteria
- [x] `vets-api` branch protection rules can be managed from code

**Comment by rjohnson2011** [comment]
Currently researching the links provided in the description (Settings app) as well as how Probot is applied to repos. Going to apply to vets-api first since that's the repo I'm most familiar with.

**Comment by rjohnson2011** [comment]
@pjhill Is it safe to assume we wish to start with the branch protections included in each repo's UI? 
For `vets-api` that would be:
1. Test
2. Linting & Securty
3. danger/danger

Just want to double check re: including these and no more / no less.

**Comment by rjohnson2011** [comment]
Looks like Github Actions has several of these branch protections tools available:
https://github.com/marketplace/automate-branch-rules
https://github.com/marketplace/actions/set-branch-protection

**Comment by pjhill** [comment]
Yeah, we would want to at least be able to reproduce the settings configured through the GH UI. I'm not sure if we want to add additional checks in the future. That's probably something to be discussed with the COP that owns each repo and @little-oddball

**Comment by laineymajor** [comment]
@pjhill Is there a small chunk of work @rjohnson2011 can start with here? Then we can revisit what's been done, make decisions, and move forward. 

If this requires discussion by the CoP, I'll have Ryan pick up a different ticket as this is his only ticket right now. 

@rjohnson2011 you and I are meeting this morning, please come to the meeting with ideas about
a) Any ideas for ways this ticket can be broken apart so it can be iterated on
b) Additional CoP tickets you can start working...

Thanks!

**Comment by rjohnson2011** [comment]
Discussed using Github Actions and modifying the settings in a .yml such as indicated here https://github.com/marketplace/actions/branch-protection with @pjhill 

That is an acceptable approach that satisfies the requirements of the ticket. 

Still TBD is discussing branch protection rules with respective codeowners.

**Comment by laineymajor** [comment]
Ryan and I discussed, he will start with BE and take an iterative approach.

**Comment by laineymajor** [comment]
@rjohnson2011 , please provide information on the actions you took yesterday with this ticket + what you plan on doing today. Thanks!

**Comment by rjohnson2011** [comment]
Created a branch [branch-protection-settings](https://github.com/department-of-veterans-affairs/vets-api/tree/branch-protection-settings) in `vets-api`. 

Extended .github workflows directory by adding a `branch_protection.yml` file. This includes adding a Github Action to pull in branch protection settings from the .github config directory. Within here the follow branch protection settings are being instantiated:
```
- branch: branch-protection-settings
  protection:
    required_status_checks:
      strict: true
      contexts:
        - lint
        - test
        - build
    required_pull_request_reviews:
      dismissal_restrictions: null
      dismiss_stale_reviews: true
      require_code_owner_reviews: false
      required_approving_review_count: 1
    required_linear_history: false
    allow_force_pushes: false
    allow_deletions: false
```

Currently working on getting this GHA to run on my branch successfully before merging to main branch. Debugging a few syntax and node version issues that have come up.

**Comment by rjohnson2011** [comment]
The Github access token does not have the full permissions required to set branch protections.

Screenshot of the Github Action run:
![image.png](https://images.zenhubusercontent.com/5fb40598d06a06671e842a63/0411e688-e0b6-4a51-91c8-3619ee95274d)

Per the documentation:
![image.png](https://images.zenhubusercontent.com/5fb40598d06a06671e842a63/9c5ff5f7-ecc7-4581-a8f0-ed807dc5ab2d)

**Comment by rjohnson2011** [comment]
Opened up an issue with the Infrastructure team - https://github.com/department-of-veterans-affairs/va.gov-team/issues/57169

**Comment by rjohnson2011** [comment]
Spoke with @pjhill regarding bringing up this issue during DevOps CoP meeting today to see if repo owner can be located to alter these permissions.

**Comment by laineymajor** [comment]
Sounds good. @rjohnson2011 let's list this ticket as blocked, and put this to the side until it's unblocked. Thanks!

**Comment by laineymajor** [comment]
@rjohnson2011 what did you find after your 2 hour search? Just making sure you don't get stuck again. Thanks!

**Comment by rjohnson2011** [comment]
@little-oddball Is it possible for you to install the Github settings app https://github.com/apps/settings for `vets-api`? As per above, someone with organization repo owner permissions is needed. If that's possible, I will open a PR with an updated CODEOWNERS file (to restrict authorized users who can modify) and update the branch protection settings with code.
CC: @pjhill

**Comment by laineymajor** [comment]
Carrying over to next sprint. Currently blocked.

**Comment by laineymajor** [comment]
@little-oddball making sure this is still on your radar... see above comments.
@rjohnson2011 please circle-back with Clint on Monday.

**Comment by rjohnson2011** [comment]
@little-oddball Just checking in once more if this is something you could help with.

**Comment by little-oddball** [comment]
@rjohnson2011 @laineymajor - I tried this and it requires a site Admin. I would suggest reaching out to Brett Logan on the #github_information Slack channel.

**Comment by rjohnson2011** [comment]
Opened a[ PR for branch protection settings](https://github.com/department-of-veterans-affairs/vets-api/pull/12631) according to the settings of the Github branch protection UI for `vets-api`.

**Comment by rjohnson2011** [comment]
Merged into Settings for `vets-api`. Currently [requested access](https://dsva.slack.com/archives/CU1E4CX9U/p1683916929961689?thread_ts=1683824202.663979&cid=CU1E4CX9U) to particular branch settings for `vets-website` and `content-build`.

**Comment by holdenhinkle** [comment]
I'm guessing this section was causing the problems today:

```
      # Required. Restrict who can push to this branch. Team and user restrictions are only available for organization-owned repositories. Set to null to disable.
      restrictions:
        apps: []
        users: []
        teams: []
```

By setting `apps`, `users`, and `teams` to an empty array, we're telling GH that we want to restrict by `apps`, `users`, and `teams` but we're not telling GH who's allowed (I think allowed `apps`, `users`, and `teams` would be elements in the arrays). We should set each to `null` per the instructions:

> Set to null to disable.

```
      # Required. Restrict who can push to this branch. Team and user restrictions are only available for organization-owned repositories. Set to null to disable.
      restrictions:
        apps: null
        users: null
        teams: null
```

**Comment by holdenhinkle** [comment]
This PR is all ready to go - https://github.com/department-of-veterans-affairs/vets-api/pull/12681

See convo here - https://dsva.slack.com/archives/C0460N83Y9G/p1684247446929079?thread_ts=1684170052.973979&cid=C0460N83Y9G

**Comment by rmtolmach** [comment]
To test https://github.com/department-of-veterans-affairs/vets-api/pull/12681, I asked the PR author of https://github.com/department-of-veterans-affairs/vets-api/pull/12766 to show me his settings before and after the config PR merged.

Before config file merged. Before getting a PR approval:
![Screenshot 2023-05-23 at 10 29 14 AM](https://github.com/department-of-veterans-affairs/va.gov-team/assets/10993987/ecf1e660-ead6-41ef-8e82-12c4d4a80c57)

After config file merged. After PR approval. 
![Screenshot 2023-05-23 at 10 30 56 AM](https://github.com/department-of-veterans-affairs/va.gov-team/assets/10993987/c79b0f51-05c2-491f-893e-efffef97c5b7)

PR author was able to merge his own PR and was not forced to update his branch–the two problems we faced last time.

**Comment by laineymajor** [comment]
@rjohnson2011 looks like this is complete, correct?

**Comment by rjohnson2011** [comment]
`settings.yml` has been updated to reflect the Github Branch Protections UI correctly -> https://github.com/department-of-veterans-affairs/vets-api/pull/12778

Worth noting is the [Branch Protection Settings UI](https://github.com/department-of-veterans-affairs/vets-api/settings/branches) is not 'locked' to the `settings.yml` file. Accordingly, the UI can be modified and overrule codified settings at any point. This limits functionality for this to be the decisive source of truth for these settings.

Currently,`settings.yml` best serves as a terrform-like infrastructure management guide for the ideal settings in the event of the UI Branch Protection Settings being unnecessarily altered / wiped out. This ticket was drawn in response to a security incident where such a event was deemed a possibility. 

A difficulty encountered is the documentation surrounding controlling branch protection settings outside Github's UI is limited, while [Github Settings Probot app solution](https://github.com/apps/settings) was deemed the best approach further discovery revealed concerns being [error prone](https://github.com/repository-settings/app/issues/625). Testing is difficult due to the complexity of repos involving particular groups permissions and extensive code checks. Each iteration must then be put through a PR and merged to master.

Open for exploration is using a Github Action to perform a scheduled deploy of this file. However, this is beyond the scope of this ticket and a heavy a lift for something rarely changed.

**Comment by holdenhinkle** [comment]
Yay!
